pkg/sink/codec/bootstraper.go

```diff
@@ -0,0 +1,263 @@
+	builder RowEventEncoderBuilder,
+	sendBootstrapInterval time.Duration,
+	sendBootstrapInMsgCount int32,
+	maxInactiveDuration time.Duration,
+) *bootstrapWorker {
+	return &bootstrapWorker{
+		changefeedID:            changefeedID,
+		outCh:                   outCh,
+		builder:                 builder,
+		activeTables:            sync.Map{},
+		sendBootstrapInterval:   sendBootstrapInterval,
+		sendBootstrapInMsgCount: sendBootstrapInMsgCount,
+		maxInactiveDuration:     maxInactiveDuration,
+	}
+}
+
+func (b *bootstrapWorker) run(ctx context.Context) error {
+	sendTicker := time.NewTicker(bootstrapWorkerTickerInterval)
+	defer sendTicker.Stop()
+	gcTicker := time.NewTicker(bootstrapWorkerGCInterval)
+	defer gcTicker.Stop()
+	errCh := make(chan error, 1)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-sendTicker.C:
+			b.activeTables.Range(func(key, value interface{}) bool {
+				table := value.(*tableStatus)
+				err := b.sendBootstrapMsg(ctx, table)
+				if err != nil {
+					errCh <- err
+					return false
+				}
+				return true
+			})
+		case <-gcTicker.C:
+			b.gcInactiveTables()
+		case err := <-errCh:
+			return err
+		}
+	}
+}
+
+func (b *bootstrapWorker) addEvent(
+	ctx context.Context,
+	key TopicPartitionKey,
+	row *model.RowChangedEvent,
+) error {
+	table, ok := b.activeTables.Load(row.Table.TableID)
+	if !ok {
+		tb := newTableStatus(key, row)
+		b.activeTables.Store(tb.id, tb)
+		// Send bootstrap message immediately when a new table is added
+		err := b.sendBootstrapMsg(ctx, tb)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	} else {
+		// If the table is already in the activeTables, update its status.
+		table.(*tableStatus).update(key, row)
+	}
+	return nil
+}
+
+// sendBootstrapMsg sends a bootstrap message if the table meets the condition
+// 1. The time since last bootstrap message sent is larger than sendBootstrapInterval
+// 2. The received row event count since last bootstrap message sent is larger than sendBootstrapInMsgCount
+// Note: It is a blocking method, it will block if the outCh is full.
+func (b *bootstrapWorker) sendBootstrapMsg(ctx context.Context, table *tableStatus) error {
+	if !table.shouldSendBootstrapMsg(
+		b.sendBootstrapInterval,
+		b.sendBootstrapInMsgCount) {
+		return nil
+	}
+	table.reset()
+	tableInfo := table.tableInfo.Load().(*model.TableInfo)
+	events, err := b.generateEvents(table.topic, table.totalPartition.Load(), tableInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, event := range events {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case b.outCh <- event:
+		}
+	}
+	return nil
+}
+
+func (b *bootstrapWorker) generateEvents(
+	topic string,
+	totalPartition int32,
+	tableInfo *model.TableInfo,
+) ([]*future, error) {
+	res := make([]*future, 0, int(totalPartition))
+	// Bootstrap messages of a table should be sent to all partitions.
+	for i := 0; i < int(totalPartition); i++ {
+		encoder := b.builder.Build()
```

Comment:
it looks that one `encoder` can be used multiple times, no need to create the encoder for each message, this allocates too many small memory

Reply:
I think the bootstrap message should be sent to the first partition.